### ISSUE TITLE
Segmentation fault when calling model.score() in python

### ISSUE THREAD

**Issue by jharaldson** [JohnLangford/vowpal_wabbit#960]
This error is seen both with Python2 and 3.
Running environment is Mac OS X 10.11.3, boost 1.6.

The last line in the below code results in "Segmentation fault: 11".
Adding a printout for each line that is run related to the last line.
[printout.txt](https://github.com/JohnLangford/vowpal_wabbit/files/176472/printout.txt)

Seems that it is possible to run ```model.score(X_train, y_train)```once, but not twice, i.e. changing the order of the last two rows does not matter.

```
# Binary Classification
from sklearn_vw import VWClassifier
import numpy as np
from sklearn import datasets
from sklearn.cross_validation import train_test_split

# get some data
X, y = datasets.make_hastie_10_2(n_samples=10000, random_state=1)
X = X.astype(np.float32)

# split train and test set
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=256)

# build vowpal wabbit model
model = VWClassifier()
model.fit(X_train, y_train)

# evaluate
print('training score: {}'.format(model.score(X_train, y_train)))
print('testing score: {}'.format(model.score(X_test, y_test)))
```

Update:
Same issue occurs when running first model.score() followed by model.predict(). Also calling model.fit() twice causes a segmentation fault.

Adding a lldb stacktrace:

```
python$ lldb python3
(lldb) target create "python3"
Current executable set to 'python3' (x86_64).
(lldb) process launch -- tmp.py
Process 1158 launched: '/usr/local/bin/python3' (x86_64)
Process 1158 stopped
* thread #1: tid = 0x9239, 0x00007fff5fc01000 dyld`_dyld_start, stop reason = exec
    frame #0: 0x00007fff5fc01000 dyld`_dyld_start
dyld`_dyld_start:
->  0x7fff5fc01000 <+0>: popq   %rdi
    0x7fff5fc01001 <+1>: pushq  $0x0
    0x7fff5fc01003 <+3>: movq   %rsp, %rbp
    0x7fff5fc01006 <+6>: andq   $-0x10, %rsp
```

**Comment by JohnLangford** [comment]
@hal3 any idea what's up?

-John

**Comment by gramhagen** [comment]
I suspect this is because score and predict methods in sklearn_vw are calling self.get_vw().finish(), I can't replicate the fault calling fit twice though.  But I'm not sure calling finish() is necessary for the other cases, @jconwell thoughts here?

**Comment by gramhagen** [comment]
This should be resolved in #972

**Comment by JohnLangford** [comment]
Closing, likely solved.  Thanks @gramhagen
